janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java

```diff
@@ -265,6 +265,13 @@
                     "performance improvement if there is a non-trivial latency to the backend.",
             ConfigOption.Type.MASKABLE, false);
 
+    public static final ConfigOption<Integer> INDEX_SELECT_BRUTE_FORCE_THRESHOLD = new ConfigOption<>(QUERY_NS, "index-select-threshold",
+            "Threshold of deciding whether to use brute force enumeration algorithm or fast approximation algorithm " +
+                    "for selecting suitable indexes. Selecting optimal indexes for a query is a NP-complete set cover problem. " +
+                    "When number of suitable index candidates is no larger than threshold, JanusGraph uses brute force search " +
+                    "with exponential time complexity to ensure the best combination of indexes is selected.",
+            ConfigOption.Type.MASKABLE, 10);
+
```

Comment:
I didn't check the code yet, but have general questions about this optimization.
1) Do I understand it correctly that if we use `index-select-threshold = 0` it means that current logic (which is currently in `master` branch) will be used? I.e., as I understand, if we use `0` it means that the behaviour will be the same as in current `master` branch and complexity of the current algorithm to select indexes will not be increased.

2) If `1)` is `yes` than I see that this value is set by default to `10`. It means that it is a breaking change. If so, we should describe this breaking change here https://github.com/JanusGraph/janusgraph/blob/master/docs/changelog.md#upgrade-instructions

3) Is it suggested to use `10` here or should user somehow decide what value to use here? Just interested, if there is 10 or less candidates, is it faster to use `brute force` algorithm or is it just sufficiently fast so that it is OK to sacrifice candidate selection performance if favour to choosing a good candidate?

Reply:
Yes, if set to 0, then the existing logic in current master branch will be used.
10 is an arbitrary number picked by me originally. My primary purpose was to pick up a threshold that is "just sufficiently fast so that it is OK to sacrifice candidate selection performance if favour to choosing a good candidate". We could make it even smaller just to be safe it will make as least impact as possible.

In practice, I think the chance is small that people have many index candidates (e.g. >= 5) for one index query.